auslib/test/test_db.py

```diff
@@ -637,6 +638,376 @@ def testHistoryGetChangeWithDataVersionWithNonPrimaryKeyColumn(self):
+        row = self.sc_table.t.select().where(self.sc_table.sc_id == 5).execute().fetchall()[0]
+        self.assertEquals(row.scheduled_by, "bob")
+        self.assertEquals(row.when, 888)
+        self.assertEquals(row.data_version, 1)
+        self.assertEquals(row.base_fooid, None)
+        self.assertEquals(row.base_foo, "newthing1")
+        self.assertEquals(row.base_bar, None)
+        self.assertEquals(row.base_data_version, None)
+
+    def testInsertWithNonAutoincrement(self):
+        class TestTable2(AUSTable):
+
+            def __init__(self, db, metadata):
+                self.table = Table("test_table2", metadata, Column("foo_name", String(15), primary_key=True),
+                                   Column("foo", String(15)),
+                                   Column("bar", String(15)))
+                super(TestTable2, self).__init__(db, "sqlite", scheduled_changes=True, history=True, versioned=True)
+
+        table = TestTable2(self.db, self.metadata)
+        self.metadata.create_all()
+        what = {"foo_name": "i'm a foo", "foo": "123", "bar": "456", "when": 876}
+        table.scheduled_changes.insert(changed_by="mary", **what)
+        row = table.scheduled_changes.t.select().where(table.scheduled_changes.sc_id == 1).execute().fetchall()[0]
+        self.assertEquals(row.scheduled_by, "mary")
+        self.assertEquals(row.when, 876)
+        self.assertEquals(row.data_version, 1)
+        self.assertEquals(row.base_foo_name, "i'm a foo")
+        self.assertEquals(row.base_foo, "123")
+        self.assertEquals(row.base_bar, "456")
+        self.assertEquals(row.base_data_version, None)
+
+    def testInsertForExistingNoSuchRow(self):
+        what = {"fooid": 10, "foo": "thing", "data_version": 1, "when": 999}
+        self.assertRaisesRegexp(ValueError, "Cannot create scheduled change with data_version for non-existent row", self.sc_table.insert, changed_by="bob",
+                                **what)
+
+    def testInsertMissingRequiredPartOfPK(self):
+        class TestTable2(AUSTable):
+
+            def __init__(self, db, metadata):
+                self.table = Table("test_table2", metadata, Column("fooid", Integer, primary_key=True, autoincrement=True),
+                                   Column("foo", String(15), primary_key=True),
+                                   Column("bar", String(15)))
+                super(TestTable2, self).__init__(db, "sqlite", scheduled_changes=True, history=True, versioned=True)
+
+        table = TestTable2("fake", self.metadata)
+        self.metadata.create_all()
+        what = {"fooid": 2, "when": 4532}
+        self.assertRaisesRegexp(ValueError, "Missing primary key column", table.scheduled_changes.insert, changed_by="bob", **what)
+
+    def testInsertWithIncompatibleConditions(self):
+        what = {"foo": "blah", "when": "abc"}
+        self.assertRaisesRegexp(ValueError, "Cannot parse", self.sc_table.insert, changed_by="bob", **what)
```

Comment:
I'm not sure what this one is testing - it appears to only have a badly formed and unused foo. The name makes me think of testValidateConditionsNotAllowedWhenAndOther().